jumanji/environments/packing/jigsaw/env.py

```diff
@@ -0,0 +1,523 @@
+            - current_board: BoundedArray (int) of shape (num_rows, num_cols).
+            - pieces: BoundedArray (int) of shape (num_pieces, 3, 3).
+            - action_mask: BoundedArray (bool) of shape (num_pieces,).
+        """
+
+        current_board = specs.BoundedArray(
+            shape=(self.num_rows, self.num_cols),
+            minimum=0,
+            maximum=self.num_pieces,
+            dtype=jnp.float32,
+            name="current_board",
+        )
+
+        pieces = specs.BoundedArray(
+            shape=(self.num_pieces, 3, 3),
+            minimum=0,
+            maximum=self.num_pieces,
+            dtype=jnp.float32,
+            name="pieces",
+        )
+
+        action_mask = specs.BoundedArray(
+            shape=(
+                self.num_pieces,
+                4,
+                self.num_rows - 2,
+                self.num_cols - 2,
+            ),
+            minimum=False,
+            maximum=True,
+            dtype=bool,
+            name="action_mask",
+        )
+
+        return specs.Spec(
+            Observation,
+            "ObservationSpec",
+            current_board=current_board,
+            pieces=pieces,
+            action_mask=action_mask,
+        )
+
+    def action_spec(self) -> specs.MultiDiscreteArray:
+        """Specifications of the action expected by the `JigSaw` environment.
+
+        Returns:
+            MultiDiscreteArray (int32) of shape (num_pieces, num_rotations,
+            max_row_position, max_col_position).
+            - num_pieces: int between 0 and num_pieces - 1 (included).
+            - num_rotations: int between 0 and 3 (included).
+            - max_row_position: int between 0 and max_row_position - 1 (included).
+            - max_col_position: int between 0 and max_col_position - 1 (included).
+        """
+
+        max_row_position = self.num_rows - 2
+        max_col_position = self.num_cols - 2
+
+        return specs.MultiDiscreteArray(
+            num_values=jnp.array(
+                [self.num_pieces, 4, max_row_position, max_col_position]
+            ),
+            name="action",
+        )
+
+    def _check_done(self, state: State) -> bool:
+        """Checks if the environment is done by checking whether the number of
+        steps is equal to the number of pieces in the puzzle.
+
+        Args:
+            state: current state of the environment.
+
+        Returns:
+            True if the environment is done, False otherwise.
+        """
+
+        done: bool = state.step_count >= state.num_pieces
+
+        return done
+
+    def _check_action_is_legal(
+        self,
+        action: chex.Numeric,
+        current_board: chex.Array,
+        placed_pieces: chex.Array,
+        grid_mask_piece: chex.Array,
+    ) -> bool:
+        """Checks if the action is legal by considering the action mask and the
+           board mask. An action is legal if the action mask is True for that action
```

Comment:
Just tab the lines that are not the first line.

Reply:
Done :)